File: day_26_string_operations/day_26_string_operations.cpp
/*
 * STRING OPERATIONS: C++17 INDUSTRY-STYLE CASE STUDY
 *
 * Scenario:
 * A lightweight command-line text processing service receives log records,
 * normalizes their text, validates fields, performs string transformations,
 * counts character and word frequencies, searches records, and produces
 * analytical output.
 *
 * The implementation demonstrates:
 * - std::string
 * - concatenation
 * - substr()
 * - insertion and deletion
 * - replacement
 * - reversal
 * - frequency counting
 * - searching
 * - validation
 * - classes and encapsulation
 * - vectors, maps, unordered_map, sets
 * - exceptions
 * - algorithmic complexity
 * - Unicode limitations in ordinary std::string
 * - production-oriented validation
 *
 * Compile:
 *   g++ -std=c++17 -O2 string_operations.cpp -o string_operations
 */

#include <algorithm>
#include <cctype>
#include <exception>
#include <iomanip>
#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;


// ---------------------------------------------------------------------------
// Utility functions
// ---------------------------------------------------------------------------

string trim(const string& text) {
    const auto first = text.find_first_not_of(" \t\r\n");
    if (first == string::npos) {
        return "";
    }

    const auto last = text.find_last_not_of(" \t\r\n");
    return text.substr(first, last - first + 1);
}

string toUpperAscii(string text) {
    for (char& character : text) {
        character = static_cast<char>(
            toupper(static_cast<unsigned char>(character))
        );
    }

    return text;
}

string normalizeWhitespace(const string& text) {
    string result;
    bool previousWasWhitespace = false;

    for (char character : text) {
        const bool whitespace =
            isspace(static_cast<unsigned char>(character)) != 0;

        if (whitespace) {
            if (!previousWasWhitespace) {
                result += ' ';
            }
        } else {
            result += character;
        }

        previousWasWhitespace = whitespace;
    }

    return trim(result);
}

string reverseString(const string& text) {
    string result = text;
    reverse(result.begin(), result.end());
    return result;
}


// ---------------------------------------------------------------------------
// Basic string operations
// ---------------------------------------------------------------------------

string concatenate(const string& first, const string& second) {
    return first + second;
}

string extractSubstring(
    const string& text,
    size_t start,
    size_t length
) {
    if (start > text.size()) {
        throw out_of_range("Substring start is outside the string.");
    }

    return text.substr(start, length);
}

string replaceAll(
    string text,
    const string& oldValue,
    const string& newValue
) {
    if (oldValue.empty()) {
        throw invalid_argument("Replacement value cannot be empty.");
    }

    size_t position = 0;

    while ((position = text.find(oldValue, position)) != string::npos) {
        text.replace(position, oldValue.size(), newValue);
        position += newValue.size();
    }

    return text;
}

string deleteAt(const string& text, size_t index) {
    if (index >= text.size()) {
        throw out_of_range("Deletion index is outside the string.");
    }

    string result = text;
    result.erase(index, 1);
    return result;
}

string insertAt(
    const string& text,
    size_t index,
    const string& value
) {
    if (index > text.size()) {
        throw out_of_range("Insertion index is outside the string.");
    }

    string result = text;
    result.insert(index, value);
    return result;
}

unordered_map<char, size_t> characterFrequency(const string& text) {
    unordered_map<char, size_t> frequency;

    for (char character : text) {
        ++frequency[character];
    }

    return frequency;
}

bool isPalindrome(string text) {
    text.erase(
        remove_if(
            text.begin(),
            text.end(),
            [](unsigned char character) {
                return !isalnum(character);
            }
        ),
        text.end()
    );

    transform(
        text.begin(),
        text.end(),
        text.begin(),
        [](unsigned char character) {
            return static_cast<char>(tolower(character));
        }
    );

    return text == string(text.rbegin(), text.rend());
}


// ---------------------------------------------------------------------------
// LogRecord
// ---------------------------------------------------------------------------

class LogRecord {
private:
    string level_;
    string user_;
    string message_;

public:
    LogRecord(
        string level,
        string user,
        string message
    )
        : level_(move(level)),
          user_(move(user)),
          message_(move(message)) {

        level_ = toUpperAscii(trim(level_));
        user_ = trim(user_);
        message_ = normalizeWhitespace(message_);

        if (level_.empty()) {
            throw invalid_argument("Log level cannot be empty.");
        }

        if (user_.empty()) {
            throw invalid_argument("User cannot be empty.");
        }

        if (message_.empty()) {
            throw invalid_argument("Message cannot be empty.");
        }
    }

    const string& level() const {
        return level_;
    }

    const string& user() const {
        return user_;
    }

    const string& message() const {
        return message_;
    }

    string serialize() const {
        return level_ + "|" + user_ + "|" + message_;
    }
};


// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

vector<string> splitByDelimiter(
    const string& text,
    char delimiter
) {
    vector<string> parts;
    string current;

    for (char character : text) {
        if (character == delimiter) {
            parts.push_back(current);
            current.clear();
        } else {
            current += character;
        }
    }

    parts.push_back(current);
    return parts;
}

bool parseLogRecord(
    const string& raw,
    LogRecord& output,
    string& error
) {
    const vector<string> parts = splitByDelimiter(raw, '|');

    if (parts.size() != 3) {
        error = "Expected exactly three fields separated by '|'.";
        return false;
    }

    try {
        output = LogRecord(parts[0], parts[1], parts[2]);
        return true;
    } catch (const exception& exception) {
        error = exception.what();
        return false;
    }
}


// ---------------------------------------------------------------------------
// Text analytics
// ---------------------------------------------------------------------------

class TextAnalytics {
public:
    static map<char, size_t> sortedCharacterFrequency(
        const string& text
    ) {
        map<char, size_t> frequency;

        for (char character : text) {
            ++frequency[character];
        }

        return frequency;
    }

    static map<string, size_t> wordFrequency(
        const string& text
    ) {
        map<string, size_t> frequency;
        string word;

        auto flushWord = [&]() {
            if (!word.empty()) {
                string normalized = toUpperAscii(word);
                ++frequency[normalized];
                word.clear();
            }
        };

        for (char character : text) {
            if (isalnum(static_cast<unsigned char>(character))) {
                word += character;
            } else {
                flushWord();
            }
        }

        flushWord();
        return frequency;
    }

    static string firstNonRepeatingCharacter(
        const string& text
    ) {
        const auto frequency = characterFrequency(text);

        for (char character : text) {
            auto iterator = frequency.find(character);

            if (iterator != frequency.end() &&
                iterator->second == 1) {
                return string(1, character);
            }
        }

        return "";
    }
};


// ---------------------------------------------------------------------------
// Production-style processor
// ---------------------------------------------------------------------------

class LogProcessor {
private:
    vector<LogRecord> validRecords_;
    vector<string> invalidRecords_;

public:
    void ingest(const vector<string>& rawRecords) {
        for (const string& raw : rawRecords) {
            LogRecord record("", "", "");
            string error;

            if (parseLogRecord(raw, record, error)) {
                validRecords_.push_back(record);
            } else {
                invalidRecords_.push_back(
                    raw + " [reason: " + error + "]"
                );
            }
        }
    }

    const vector<LogRecord>& records() const {
        return validRecords_;
    }

    const vector<string>& invalidRecords() const {
        return invalidRecords_;
    }

    map<string, size_t> levelFrequency() const {
        map<string, size_t> result;

        for (const auto& record : validRecords_) {
            ++result[record.level()];
        }

        return result;
    }

    map<string, size_t> userFrequency() const {
        map<string, size_t> result;

        for (const auto& record : validRecords_) {
            ++result[record.user()];
        }

        return result;
    }

    vector<LogRecord> findContaining(
        const string& searchTerm
    ) const {
        vector<LogRecord> result;

        for (const auto& record : validRecords_) {
            if (
                record.message().find(searchTerm) != string::npos ||
                record.user().find(searchTerm) != string::npos
            ) {
                result.push_back(record);
            }
        }

        return result;
    }

    string combinedMessages() const {
        string result;

        for (size_t index = 0; index < validRecords_.size(); ++index) {
            if (index > 0) {
                result += '\n';
            }

            result += validRecords_[index].message();
        }

        return result;
    }
};


// ---------------------------------------------------------------------------
// Output helpers
// ---------------------------------------------------------------------------

void printMap(const map<string, size_t>& values) {
    for (const auto& [key, value] : values) {
        cout << "  " << key << ": " << value << '\n';
    }
}

void printCharacterMap(
    const map<char, size_t>& values
) {
    for (const auto& [key, value] : values) {
        if (key == ' ') {
            cout << "  [space]: " << value << '\n';
        } else {
            cout << "  [" << key << "]: " << value << '\n';
        }
    }
}

void printRecords(
    const vector<LogRecord>& records
) {
    for (const auto& record : records) {
        cout << "  " << record.serialize() << '\n';
    }
}


// ---------------------------------------------------------------------------
// Demonstrations
// ---------------------------------------------------------------------------

void basicOperationsDemo() {
    cout << "\n" << string(70, '=') << '\n';
    cout << "1. BASIC STRING OPERATIONS\n";
    cout << string(70, '=') << '\n';

    const string first = "String";
    const string second = "Operations";

    cout << "Concatenation: "
         << concatenate(first, " " + second)
         << '\n';

    const string text = "Programming";

    cout << "Substring: "
         << extractSubstring(text, 0, 7)
         << '\n';

    cout << "Reverse: "
         << reverseString(text)
         << '\n';

    cout << "Replacement: "
         << replaceAll(text, "Programming", "C++")
         << '\n';

    cout << "Deletion: "
         << deleteAt(text, 3)
         << '\n';

    cout << "Insertion: "
         << insertAt(text, 11, "!")
         << '\n';

    cout << "Frequency of characters:\n";

    const auto frequency = characterFrequency("banana");

    map<char, size_t> sortedFrequency(
        frequency.begin(),
        frequency.end()
    );

    printCharacterMap(sortedFrequency);
}


void edgeCasesDemo() {
    cout << "\n" << string(70, '=') << '\n';
    cout << "2. EDGE CASES AND ERROR HANDLING\n";
    cout << string(70, '=') << '\n';

    try {
        cout << "Invalid substring: "
             << extractSubstring("abc", 20, 2)
             << '\n';
    } catch (const exception& exception) {
        cout << "Handled substring error: "
             << exception.what()
             << '\n';
    }

    try {
        cout << "Invalid deletion: "
             << deleteAt("abc", 20)
             << '\n';
    } catch (const exception& exception) {
        cout << "Handled deletion error: "
             << exception.what()
             << '\n';
    }

    try {
        cout << "Invalid replacement: "
             << replaceAll("abc", "", "x")
             << '\n';
    } catch (const exception& exception) {
        cout << "Handled replacement error: "
             << exception.what()
             << '\n';
    }

    cout << "Empty string length: "
         << string("").size()
         << '\n';

    cout << "Empty search result: "
         << ("abc".find("") == 0)
         << '\n';
}


void caseStudyDemo() {
    cout << "\n" << string(70, '=') << '\n';
    cout << "3. LOG PROCESSING CASE STUDY\n";
    cout << string(70, '=') << '\n';

    const vector<string> rawRecords = {
        "INFO|alice|Login successful",
        "ERROR|bob|Invalid password",
        "INFO|alice|Viewed dashboard",
        "WARNING|charlie|Password expires soon",
        "ERROR|bob|Account temporarily locked",
        "INVALID RECORD",
        "ERROR||Missing user",
        "INFO|david|   Excessive   whitespace   "
    };

    LogProcessor processor;
    processor.ingest(rawRecords);

    cout << "\nValid records:\n";
    printRecords(processor.records());

    cout << "\nInvalid records:\n";
    for (const auto& record : processor.invalidRecords()) {
        cout << "  " << record << '\n';
    }

    cout << "\nLevel frequency:\n";
    printMap(processor.levelFrequency());

    cout << "\nUser frequency:\n";
    printMap(processor.userFrequency());

    cout << "\nRecords containing 'password':\n";
    printRecords(processor.findContaining("password"));

    const string combined = processor.combinedMessages();

    cout << "\nCombined message palindrome: "
         << boolalpha
         << isPalindrome(combined)
         << '\n';

    cout << "\nWord frequency:\n";
    printMap(TextAnalytics::wordFrequency(combined));

    cout << "\nFirst non-repeating character: ";

    const string firstUnique =
        TextAnalytics::firstNonRepeatingCharacter(combined);

    if (firstUnique.empty()) {
        cout << "none\n";
    } else {
        cout << firstUnique << '\n';
    }
}


// ---------------------------------------------------------------------------
// Complexity notes encoded as runtime demonstrations
// ---------------------------------------------------------------------------

void algorithmDemo() {
    cout << "\n" << string(70, '=') << '\n';
    cout << "4. ALGORITHMIC CONSIDERATIONS\n";
    cout << string(70, '=') << '\n';

    const string text = "banana";

    /*
     * Character frequency:
     *   Time: O(n)
     *   Auxiliary space: O(k), where k is the number of distinct characters.
     *
     * Substring extraction:
     *   In modern C++, std::string::substr creates a new string, so copying
     *   the selected characters costs O(m), where m is substring length.
     *
     * Searching with find():
     *   The exact complexity depends on the standard-library implementation.
     *   It should not automatically be treated as constant time.
     */

    cout << "Text: " << text << '\n';

    const auto frequency = TextAnalytics::sortedCharacterFrequency(text);

    cout << "Frequency:\n";
    printCharacterMap(frequency);

    cout << "Palindrome 'level': "
         << boolalpha
         << isPalindrome("level")
         << '\n';
}


// ---------------------------------------------------------------------------
// Unicode considerations
// ---------------------------------------------------------------------------

void unicodeDemo() {
    cout << "\n" << string(70, '=') << '\n';
    cout << "5. CHARACTER ENCODING CONSIDERATION\n";
    cout << string(70, '=') << '\n';

    /*
     * std::string is a sequence of bytes, not automatically a sequence of
     * Unicode characters.
     *
     * ASCII examples work naturally because one ASCII character is one byte.
     * UTF-8 text may use multiple bytes for one user-perceived character.
     *
     * This program therefore treats std::string as byte-oriented text.
     * Production multilingual systems should explicitly choose and enforce
     * an encoding strategy rather than assuming byte count equals character
     * count.
     */

    const string ascii = "cafe";

    cout << "ASCII text: " << ascii << '\n';
    cout << "Byte count: " << ascii.size() << '\n';

    const string utf8 = "cafe with UTF-8: café";

    cout << "UTF-8 text: " << utf8 << '\n';
    cout << "std::string byte count: " << utf8.size() << '\n';
}


// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

void runTests() {
    cout << "\n" << string(70, '=') << '\n';
    cout << "6. SELF-TESTS\n";
    cout << string(70, '=') << '\n';

    if (concatenate("Hello", " World") != "Hello World") {
        throw runtime_error("Concatenation test failed.");
    }

    if (extractSubstring("abcdef", 1, 3) != "bcd") {
        throw runtime_error("Substring test failed.");
    }

    if (reverseString("abcdef") != "fedcba") {
        throw runtime_error("Reverse test failed.");
    }

    if (replaceAll("banana", "a", "o") != "bonono") {
        throw runtime_error("Replacement test failed.");
    }

    if (deleteAt("banana", 2) != "baana") {
        throw runtime_error("Deletion test failed.");
    }

    if (insertAt("Pythn", 4, "o") != "Python") {
        throw runtime_error("Insertion test failed.");
    }

    if (characterFrequency("banana").at('a') != 3) {
        throw runtime_error("Frequency test failed.");
    }

    if (!isPalindrome("A man, a plan, a canal: Panama")) {
        throw runtime_error("Palindrome test failed.");
    }

    cout << "All tests passed.\n";
}


// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

int main() {
    try {
        basicOperationsDemo();
        edgeCasesDemo();
        caseStudyDemo();
        algorithmDemo();
        unicodeDemo();
        runTests();

        cout << "\n" << string(70, '=') << '\n';
        cout << "C++ STRING OPERATIONS CASE STUDY COMPLETE\n";
        cout << string(70, '=') << '\n';

        return 0;
    } catch (const exception& exception) {
        cerr << "Fatal error: " << exception.what() << '\n';
        return 1;
    }
}

File: day_26_string_operations/day_26_string_operations.js
/*
 * STRING OPERATIONS: JAVASCRIPT PRACTICAL STUDY FILE
 *
 * Demonstrates:
 * - Creation and indexing
 * - Concatenation
 * - Substrings and slices
 * - Reversal
 * - Replacement
 * - Deletion and insertion
 * - Frequency counting
 * - Searching
 * - Unicode
 * - Regular expressions
 * - Validation
 * - Immutability
 * - Functional processing
 * - Performance-aware construction
 * - A practical text-analysis case study
 *
 * Run with:
 *   node string_operations.js
 */

// ---------------------------------------------------------------------------
// 1. FUNDAMENTALS
// ---------------------------------------------------------------------------

function fundamentals() {
    console.log("\n" + "=".repeat(70));
    console.log("1. STRING FUNDAMENTALS");
    console.log("=".repeat(70));

    const text = "JavaScript";

    console.log("String:", text);
    console.log("Type:", typeof text);
    console.log("Length:", text.length);
    console.log("First character:", text[0]);
    console.log("Last character:", text[text.length - 1]);

    for (let index = 0; index < text.length; index++) {
        console.log(`index=${index}: ${text[index]}`);
    }

    // Strings are immutable. Operations create new strings.
    const changed = "J" + text.slice(1);
    console.log("Reconstructed string:", changed);
}


// ---------------------------------------------------------------------------
// 2. CONCATENATION
// ---------------------------------------------------------------------------

function concatenationExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("2. CONCATENATION");
    console.log("=".repeat(70));

    const firstName = "Atul";
    const lastName = "Pandey";

    console.log("Using +:", firstName + " " + lastName);

    const age = 30;
    console.log(`Using template literal: ${firstName} ${lastName} is ${age}.`);

    const words = ["JavaScript", "String", "Operations"];
    console.log("Using join():", words.join(" "));

    console.log("Repetition:", "ab".repeat(3));
}


// ---------------------------------------------------------------------------
// 3. SUBSTRING EXTRACTION
// ---------------------------------------------------------------------------

function substringExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("3. SUBSTRING EXTRACTION");
    console.log("=".repeat(70));

    const text = "String Operations";

    console.log("Original:", text);
    console.log("First six:", text.slice(0, 6));
    console.log("From index 7:", text.slice(7));
    console.log("Last five:", text.slice(-5));
    console.log("Middle:", text.substring(7, 17));

    // slice() supports negative indexes; substring() treats negative values
    // differently, so slice() is usually easier for index-based extraction.
    console.log("Reverse:", [...text].reverse().join(""));
}


// ---------------------------------------------------------------------------
// 4. REVERSING
// ---------------------------------------------------------------------------

function reverseString(text) {
    // Spread into an array first. This handles many Unicode code points better
    // than text.split("") because some Unicode characters occupy multiple
    // UTF-16 code units.
    return [...text].reverse().join("");
}

function reverseExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("4. REVERSING");
    console.log("=".repeat(70));

    const text = "algorithm";

    console.log("Original:", text);
    console.log("Reversed:", reverseString(text));

    const candidate = "level";
    console.log("Palindrome:", candidate === reverseString(candidate));
}


// ---------------------------------------------------------------------------
// 5. CHARACTER REPLACEMENT
// ---------------------------------------------------------------------------

function replacementExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("5. CHARACTER REPLACEMENT");
    console.log("=".repeat(70));

    const text = "banana";

    console.log("Original:", text);
    console.log("First 'a':", text.replace("a", "o"));
    console.log("All 'a':", text.replaceAll("a", "o"));

    const sentence = "JavaScript is powerful. JavaScript runs in browsers.";
    console.log(sentence.replaceAll("JavaScript", "JS"));
}


// ---------------------------------------------------------------------------
// 6. CHARACTER DELETION
// ---------------------------------------------------------------------------

function deleteAt(text, index) {
    if (!Number.isInteger(index) || index < 0 || index >= text.length) {
        throw new RangeError("Deletion index is outside the valid range.");
    }

    return text.slice(0, index) + text.slice(index + 1);
}

function deleteRange(text, start, end) {
    if (
        !Number.isInteger(start) ||
        !Number.isInteger(end) ||
        start < 0 ||
        end < start ||
        end > text.length
    ) {
        throw new RangeError("Invalid deletion range.");
    }

    return text.slice(0, start) + text.slice(end);
}

function deletionExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("6. CHARACTER DELETION");
    console.log("=".repeat(70));

    console.log("Delete index 2:", deleteAt("banana", 2));
    console.log("Delete range:", deleteRange("JavaScript", 4, 10));
    console.log("Delete all spaces:", "a b c".replaceAll(" ", ""));

    try {
        deleteAt("abc", 99);
    } catch (error) {
        console.log("Expected error:", error.message);
    }
}


// ---------------------------------------------------------------------------
// 7. CHARACTER INSERTION
// ---------------------------------------------------------------------------

function insertAt(text, index, value) {
    if (!Number.isInteger(index) || index < 0 || index > text.length) {
        throw new RangeError("Insertion index is outside the valid range.");
    }

    return text.slice(0, index) + value + text.slice(index);
}

function insertionExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("7. CHARACTER INSERTION");
    console.log("=".repeat(70));

    console.log("Correction:", insertAt("Javasript", 4, "c"));
    console.log("Beginning:", insertAt("World", 0, "Hello "));
    console.log("End:", insertAt("Hello", 5, "!"));
}


// ---------------------------------------------------------------------------
// 8. FREQUENCY COUNTING
// ---------------------------------------------------------------------------

function characterFrequency(text) {
    const frequency = new Map();

    for (const character of text) {
        frequency.set(character, (frequency.get(character) || 0) + 1);
    }

    return frequency;
}

function wordFrequency(text) {
    const words = text.toLocaleLowerCase().match(/\b[\p{L}\p{N}'-]+\b/gu) || [];
    const frequency = new Map();

    for (const word of words) {
        frequency.set(word, (frequency.get(word) || 0) + 1);
    }

    return frequency;
}

function mapToObject(map) {
    return Object.fromEntries(map.entries());
}

function frequencyExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("8. FREQUENCY COUNTING");
    console.log("=".repeat(70));

    console.log(
        "Character frequency:",
        mapToObject(characterFrequency("banana"))
    );

    const sentence = "JavaScript is useful and JavaScript is flexible";
    console.log("Word frequency:", mapToObject(wordFrequency(sentence)));
}


// ---------------------------------------------------------------------------
// 9. SEARCHING
// ---------------------------------------------------------------------------

function searchingExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("9. SEARCHING");
    console.log("=".repeat(70));

    const text = "String operations are useful.";

    console.log("Includes:", text.includes("operations"));
    console.log("Index:", text.indexOf("operations"));
    console.log("Missing index:", text.indexOf("database"));
    console.log("Starts with:", text.startsWith("String"));
    console.log("Ends with:", text.endsWith("."));
}


// ---------------------------------------------------------------------------
// 10. NORMALIZATION
// ---------------------------------------------------------------------------

function normalizationExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("10. NORMALIZATION");
    console.log("=".repeat(70));

    const text = "   JavaScript STRING Operations   ";

    console.log("Original:", JSON.stringify(text));
    console.log("Trimmed:", JSON.stringify(text.trim()));
    console.log("Lowercase:", text.toLowerCase());
    console.log("Uppercase:", text.toUpperCase());

    const userInput = "  YES  ";
    const normalized = userInput.trim().toLowerCase();

    console.log("Normalized input:", normalized);
    console.log("Accepted:", ["yes", "y"].includes(normalized));
}


// ---------------------------------------------------------------------------
// 11. REGULAR EXPRESSIONS
// ---------------------------------------------------------------------------

function regexExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("11. REGULAR EXPRESSIONS");
    console.log("=".repeat(70));

    const text = "Contact alice@example.com and bob@example.org.";

    const emails = text.match(
        /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g
    ) || [];

    console.log("Emails:", emails);

    const messy = "JavaScript\tstrings   are\npowerful";
    const clean = messy.replace(/\s+/g, " ").trim();

    console.log("Normalized whitespace:", clean);
}


// ---------------------------------------------------------------------------
// 12. UNICODE
// ---------------------------------------------------------------------------

function unicodeExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("12. UNICODE AND UTF-16");
    console.log("=".repeat(70));

    const examples = ["A", "é", "中", "🙂"];

    for (const character of examples) {
        console.log(
            character,
            "UTF-16 length:",
            character.length,
            "code points:",
            [...character].length
        );
    }

    const multilingual = "नमस्ते 世界";
    console.log("Multilingual text:", multilingual);
    console.log("UTF-16 code-unit length:", multilingual.length);
    console.log("Code-point count:", [...multilingual].length);
}


// ---------------------------------------------------------------------------
// 13. VALIDATION
// ---------------------------------------------------------------------------

function validateIdentifier(identifier) {
    if (identifier.length === 0) {
        return { valid: false, message: "Identifier cannot be empty." };
    }

    if (!/^[A-Za-z_$]/.test(identifier)) {
        return {
            valid: false,
            message: "Identifier must begin with a letter, _ or $."
        };
    }

    if (!/^[A-Za-z0-9_$]+$/.test(identifier)) {
        return {
            valid: false,
            message: "Identifier contains an invalid character."
        };
    }

    return { valid: true, message: "Valid identifier." };
}

function validationExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("13. VALIDATION");
    console.log("=".repeat(70));

    const candidates = [
        "total_count",
        "$price",
        "2ndValue",
        "user-name",
        "",
        "user123"
    ];

    for (const candidate of candidates) {
        console.log(candidate, validateIdentifier(candidate));
    }
}


// ---------------------------------------------------------------------------
// 14. TEXT ANALYZER CLASS
// ---------------------------------------------------------------------------

class TextAnalyzer {
    constructor(text) {
        if (typeof text !== "string") {
            throw new TypeError("TextAnalyzer requires a string.");
        }

        this.text = text;
    }

    characterFrequency(ignoreWhitespace = false) {
        const source = ignoreWhitespace
            ? [...this.text].filter(character => !/\s/u.test(character)).join("")
            : this.text;

        return characterFrequency(source);
    }

    words() {
        return this.text.match(/\b[\p{L}\p{N}'-]+\b/gu) || [];
    }

    wordFrequency() {
        return wordFrequency(this.text);
    }

    reverse() {
        return reverseString(this.text);
    }

    palindrome() {
        const normalized = [...this.text]
            .filter(character => /[\p{L}\p{N}]/u.test(character))
            .join("")
            .toLocaleLowerCase();

        return normalized === reverseString(normalized);
    }

    replace(oldValue, newValue) {
        return this.text.replaceAll(oldValue, newValue);
    }

    insert(index, value) {
        return insertAt(this.text, index, value);
    }

    deleteRange(start, end) {
        return deleteRange(this.text, start, end);
    }
}

function analyzerDemo() {
    console.log("\n" + "=".repeat(70));
    console.log("14. TEXT ANALYZER");
    console.log("=".repeat(70));

    const analyzer = new TextAnalyzer(
        "JavaScript strings are immutable. JavaScript strings are sequences."
    );

    console.log(
        "Character frequency:",
        mapToObject(analyzer.characterFrequency(true))
    );
    console.log("Word frequency:", mapToObject(analyzer.wordFrequency()));
    console.log("Reversed:", analyzer.reverse());
    console.log("Palindrome:", analyzer.palindrome());
    console.log("Replacement:", analyzer.replace("JavaScript", "JS"));
    console.log("Insertion:", analyzer.insert(0, "[TEXT] "));
    console.log("Deletion:", analyzer.deleteRange(0, 11));
}


// ---------------------------------------------------------------------------
// 15. ALGORITHMIC OPERATIONS
// ---------------------------------------------------------------------------

function firstNonRepeatingCharacter(text) {
    const frequencies = characterFrequency(text);

    for (const character of text) {
        if (frequencies.get(character) === 1) {
            return character;
        }
    }

    return null;
}

function isPalindrome(text) {
    const normalized = [...text]
        .filter(character => /[\p{L}\p{N}]/u.test(character))
        .join("")
        .toLocaleLowerCase();

    return normalized === reverseString(normalized);
}

function algorithmExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("15. ALGORITHMS");
    console.log("=".repeat(70));

    for (const value of [
        "level",
        "A man, a plan, a canal: Panama",
        "JavaScript"
    ]) {
        console.log(value, "palindrome:", isPalindrome(value));
    }

    for (const value of ["swiss", "aabbcc", ""]) {
        console.log(
            value,
            "first non-repeating:",
            firstNonRepeatingCharacter(value)
        );
    }
}


// ---------------------------------------------------------------------------
// 16. PRACTICAL LOG PROCESSING CASE STUDY
// ---------------------------------------------------------------------------

function processLogRecords(records) {
    const levelFrequency = new Map();
    const userFrequency = new Map();
    const messages = [];
    const invalidRecords = [];

    for (const record of records) {
        const parts = record.split("|");

        if (parts.length !== 3) {
            invalidRecords.push(record);
            continue;
        }

        const [rawLevel, rawUser, rawMessage] = parts;
        const level = rawLevel.trim().toUpperCase();
        const user = rawUser.trim();
        const message = rawMessage.trim().replace(/\s+/g, " ");

        if (!level || !user || !message) {
            invalidRecords.push(record);
            continue;
        }

        levelFrequency.set(
            level,
            (levelFrequency.get(level) || 0) + 1
        );

        userFrequency.set(
            user,
            (userFrequency.get(user) || 0) + 1
        );

        messages.push(message);
    }

    return {
        levels: mapToObject(levelFrequency),
        users: mapToObject(userFrequency),
        messages,
        invalid: invalidRecords
    };
}

function caseStudyDemo() {
    console.log("\n" + "=".repeat(70));
    console.log("16. LOG PROCESSING CASE STUDY");
    console.log("=".repeat(70));

    const records = [
        "INFO|alice|Login successful",
        "ERROR|bob|Invalid password",
        "INFO|alice|Viewed dashboard",
        "WARNING|charlie|Password expires soon",
        "ERROR|bob|Account temporarily locked",
        "INVALID RECORD"
    ];

    console.log(processLogRecords(records));
}


// ---------------------------------------------------------------------------
// 17. PERFORMANCE-AWARE CONSTRUCTION
// ---------------------------------------------------------------------------

function performanceExample() {
    console.log("\n" + "=".repeat(70));
    console.log("17. PERFORMANCE-AWARE STRING CONSTRUCTION");
    console.log("=".repeat(70));

    // Collect pieces and join once instead of repeatedly constructing a
    // growing string inside a large loop.
    const pieces = [];

    for (let index = 0; index < 10000; index++) {
        pieces.push(`record-${index}`);
    }

    const result = pieces.join("\n");

    console.log("Records:", pieces.length);
    console.log("Generated characters:", result.length);
}


// ---------------------------------------------------------------------------
// 18. ERROR HANDLING AND EDGE CASES
// ---------------------------------------------------------------------------

function edgeCaseExamples() {
    console.log("\n" + "=".repeat(70));
    console.log("18. EDGE CASES");
    console.log("=".repeat(70));

    const cases = ["", " ", "A", "AAAA", "12345", "😀😀"];

    for (const value of cases) {
        console.log({
            value,
            length: value.length,
            codePoints: [...value].length,
            reversed: reverseString(value),
            aCount: [...value].filter(c => c === "a").length
        });
    }

    // indexOf returns -1 rather than throwing when a value is absent.
    console.log("Missing search:", "abc".indexOf("z"));

    // slice safely handles indexes beyond the string's length.
    console.log("Out-of-range slice:", "abc".slice(100));

    try {
        insertAt("abc", 10, "X");
    } catch (error) {
        console.log("Insertion failure:", error.message);
    }
}


// ---------------------------------------------------------------------------
// 19. TESTS
// ---------------------------------------------------------------------------

function runTests() {
    console.log("\n" + "=".repeat(70));
    console.log("19. SELF-TESTS");
    console.log("=".repeat(70));

    console.assert("Hello" + " " + "World" === "Hello World");
    console.assert("abcdef".slice(1, 4) === "bcd");
    console.assert(reverseString("abcdef") === "fedcba");
    console.assert("banana".replaceAll("a", "o") === "bonono");
    console.assert(deleteAt("banana", 2) === "baana");
    console.assert(insertAt("Pythn", 4, "o") === "Python");
    console.assert(characterFrequency("banana").get("a") === 3);
    console.assert(isPalindrome("Race car!"));
    console.assert(firstNonRepeatingCharacter("swiss") === "w");
    console.assert(validateIdentifier("user_123").valid);
    console.assert(!validateIdentifier("123user").valid);

    console.log("All tests passed.");
}


// ---------------------------------------------------------------------------
// 20. MAIN
// ---------------------------------------------------------------------------

function main() {
    fundamentals();
    concatenationExamples();
    substringExamples();
    reverseExamples();
    replacementExamples();
    deletionExamples();
    insertionExamples();
    frequencyExamples();
    searchingExamples();
    normalizationExamples();
    regexExamples();
    unicodeExamples();
    validationExamples();
    analyzerDemo();
    algorithmExamples();
    caseStudyDemo();
    performanceExample();
    edgeCaseExamples();
    runTests();

    console.log("\n" + "=".repeat(70));
    console.log("JAVASCRIPT STRING OPERATIONS STUDY PROGRAM COMPLETE");
    console.log("=".repeat(70));
}

main();
